deps/patches/gmp_alloc_overflow_func.patch

```diff
@@ -0,0 +1,230 @@
++__GMP_DECLSPEC void mp_set_alloc_overflow_function (void (*) (void)) __GMP_NOTHROW;
++
++#define mp_get_alloc_overflow_function __gmp_get_alloc_overflow_function
++__GMP_DECLSPEC void mp_get_alloc_overflow_function (void (**) (void)) __GMP_NOTHROW;
++
++
+ #define mp_bits_per_limb __gmp_bits_per_limb
+ __GMP_DECLSPEC extern const int mp_bits_per_limb;
+ 
+ #define gmp_errno __gmp_errno
+ __GMP_DECLSPEC extern int gmp_errno;
+diff --git a/gmp-impl.h b/gmp-impl.h
+--- a/gmp-impl.h
++++ b/gmp-impl.h
+@@ -694,14 +694,16 @@ struct tmp_debug_entry_t {
+ /* Enhancement: __gmp_allocate_func could have "__attribute__ ((malloc))",
+    but current gcc (3.0) doesn't seem to support that.  */
+ __GMP_DECLSPEC extern void * (*__gmp_allocate_func) (size_t);
+ __GMP_DECLSPEC extern void * (*__gmp_reallocate_func) (void *, size_t, size_t);
+ __GMP_DECLSPEC extern void   (*__gmp_free_func) (void *, size_t);
++__GMP_DECLSPEC extern void   (*__gmp_alloc_overflow_func)(void);
+ 
+ __GMP_DECLSPEC void *__gmp_default_allocate (size_t);
+ __GMP_DECLSPEC void *__gmp_default_reallocate (void *, size_t, size_t);
+ __GMP_DECLSPEC void __gmp_default_free (void *, size_t);
++__GMP_DECLSPEC void __gmp_default_alloc_overflow (void);
+ 
+ #define __GMP_ALLOCATE_FUNC_TYPE(n,type) \
+   ((type *) (*__gmp_allocate_func) ((n) * sizeof (type)))
+ #define __GMP_ALLOCATE_FUNC_LIMBS(n)   __GMP_ALLOCATE_FUNC_TYPE (n, mp_limb_t)
+ 
+@@ -725,10 +727,16 @@ struct tmp_debug_entry_t {
+     if ((oldsize) != (newsize))						\
+       (ptr) = (type *) (*__gmp_reallocate_func)				\
+ 	(ptr, (oldsize) * sizeof (type), (newsize) * sizeof (type));	\
+   } while (0)
+ 
++#define __GMP_ALLOC_OVERFLOW_FUNC()                              \
++  do {                                                           \
++    (*__gmp_alloc_overflow_func) ();                             \
++    fprintf (stderr, "unexpected return from alloc_overflow\n"); \
++    abort ();                                                    \
++  } while (0)
+ 
+ /* Dummy for non-gcc, code involving it will go dead. */
+ #if ! defined (__GNUC__) || __GNUC__ < 2
+ #define __builtin_constant_p(x)   0
+ #endif
+diff --git a/memory.c b/memory.c
+--- a/memory.c
++++ b/memory.c
+@@ -36,10 +36,11 @@ see https://www.gnu.org/licenses/.  */
+ 
+ 
+ void * (*__gmp_allocate_func) (size_t) = __gmp_default_allocate;
+ void * (*__gmp_reallocate_func) (void *, size_t, size_t) = __gmp_default_reallocate;
+ void   (*__gmp_free_func) (void *, size_t) = __gmp_default_free;
++void   (*__gmp_alloc_overflow_func) () = __gmp_default_alloc_overflow;
+ 
+ 
+ /* Default allocation functions.  In case of failure to allocate/reallocate
+    an error message is written to stderr and the program aborts.  */
+ 
+@@ -142,5 +143,12 @@ void
+     blk_ptr = p - 1;
+   }
+ #endif
+   free (blk_ptr);
+ }
++
++void
++__gmp_default_alloc_overflow(void)
++{
++    fprintf(stderr, "__gmp_default_alloc_overflow called\n");
```

Comment:
This print doesn't seem necessary --- probably best to preserve the old behavior by default.

Reply:
Ack